### PR TITLE
Read only queries

### PR DESCRIPTION
**Pull request by nikitagromov**

### CODE REVIEW

File: nefertari_sqla/documents.py [hunk 1]
```diff
@@ -1069,6 +1069,67 @@ def _is_created(self):
...
+
+        if pkey_value:
+            return cls._wrap_read_only(
+                cls.session_factory().query(cls).filter_by(**{cls.pk_field(): pkey_value}).values(*columns)
+            )
+        if params:
+            return cls._wrap_read_only(
+                cls.session_factory().query(cls).filter_by(**params).values(*columns)
+            )
+        raise NoResultFound('There is no results according to passed params')
```
Comment (nmenardg) [inline nefertari_sqla/documents.py]: This is not if there's no result, it's if no params were given.

File: nefertari_sqla/documents.py [hunk 2]
```diff
@@ -1069,6 +1069,67 @@ def _is_created(self):
...
+
+    @classmethod
+    def get_columns(cls):
+        inst = inspect(cls)
```
Comment (nmenardg) [inline nefertari_sqla/documents.py]: What's `inst`?

File: nefertari_sqla/documents.py [hunk 3]
```diff
@@ -1069,6 +1069,67 @@ def _is_created(self):
...
+        return [c_attr.key for c_attr in inst.mapper.columns]
+
+    @classmethod
+    def _wrap_read_only(cls, gen):
+        read_only_cls = cls._get_readonly_cls()
+        packed_value = tuple(gen)
+        if packed_value:
+            single_value, *_ = packed_value
```
Comment (nmenardg) [inline nefertari_sqla/documents.py]: Wouldn't `single_value = next(gen, None)` do the same thing?

File: nefertari_sqla/documents.py [hunk 4]
```diff
@@ -1069,6 +1069,67 @@ def _is_created(self):
...
+            return read_only_cls(item=single_value)
+        return None
+
+    @classmethod
+    def get_readonly_collection(cls, **params):
+        if params:
+            query = cls.session_factory().query(cls).filter_by(**params)
+        else:
+            statement = text('SELECT * FROM public.%s' % cls.__tablename__)
+            query = cls.session_factory().query(cls).from_statement(statement)
```
Comment (nmenardg) [inline nefertari_sqla/documents.py:1106]: Wouldn't this work?
```python
cls.session_factory().query(cls)
```

**Review by nmenardg — CHANGES_REQUESTED** [review]

File: nefertari_sqla/documents.py (hunk 4 above)
Comment (nikitagromov) [reply nefertari_sqla/documents.py:1106]: @nmenardg you mean without `from_statement` ?

File: nefertari_sqla/documents.py (hunk 1 above)
Comment (nikitagromov) [reply nefertari_sqla/documents.py]: @nmenardg what do you think which kind of exception we should raise here ?

**Review by nmenardg — APPROVED** [review]

**Pull request merged**